packages/database/database.go

```diff
@@ -1,134 +1,65 @@
-	})
-	return err
-}
-
-func (this *prefixDb) Set(key []byte, value []byte) error {
-	e := badger.NewEntry(append(this.prefix, key...), value)
-	return this.setEntry(e)
-}
-
-func (this *prefixDb) SetWithTTL(key []byte, value []byte, ttl time.Duration) error {
-	e := badger.NewEntry(append(this.prefix, key...), value).WithTTL(ttl)
-	return this.setEntry(e)
-}
-
-func (this *prefixDb) Contains(key []byte) (bool, error) {
-	err := this.db.View(func(txn *badger.Txn) error {
-		_, err := txn.Get(append(this.prefix, key...))
-		return err
-	})
-
-	if err == ErrKeyNotFound {
-		return false, nil
-	} else {
-		return err == nil, err
-	}
-}
-
-func (this *prefixDb) Get(key []byte) ([]byte, error) {
-	var result []byte = nil
-
-	err := this.db.View(func(txn *badger.Txn) error {
-		item, err := txn.Get(append(this.prefix, key...))
+		var dbDirClear bool
+		// check whether the database is new, by checking whether any file exists within
+		// the database directory
+		fileInfos, err := ioutil.ReadDir(dbDir)
 		if err != nil {
-			return err
+			// panic on other errors, for example permission related
+			if !os.IsNotExist(err) {
+				panic(err)
+			}
+			dbDirClear = true
+		}
+		if len(fileInfos) == 0 {
+			dbDirClear = true
 		}
 
-		return item.Value(func(val []byte) error {
-			result = append([]byte{}, val...)
-
-			return nil
-		})
-	})
-
-	return result, err
-}
-
-func (this *prefixDb) Delete(key []byte) error {
-	err := this.db.Update(func(txn *badger.Txn) error {
-		return txn.Delete(append(this.prefix, key...))
-	})
-	return err
-}
-
-func (this *prefixDb) forEach(prefix []byte, consumer func([]byte, []byte)) error {
-	err := this.db.View(func(txn *badger.Txn) error {
-		iteratorOptions := badger.DefaultIteratorOptions
-		iteratorOptions.Prefix = prefix // filter by prefix
-
-		// create an iterator the default options
-		it := txn.NewIterator(iteratorOptions)
-		defer it.Close()
-
-		// loop through every key-value-pair and call the function
-		for it.Rewind(); it.Valid(); it.Next() {
-			item := it.Item()
-
-			value, err := item.ValueCopy(nil)
-			if err != nil {
-				return err
-			}
+		opts := badger.DefaultOptions(dbDir)
+		opts.Logger = nil
+		if runtime.GOOS == "windows" {
+			opts = opts.WithTruncate(true)
+		}
 
-			consumer(item.Key()[len(this.prefix):], value)
+		db, err := database.CreateDB(dbDir, opts)
+		if err != nil {
+			// errors should cause a panic to avoid singleton deadlocks
+			panic(err)
 		}
-		return nil
-	})
-	return err
-}
+		instance = db
 
-func (this *prefixDb) ForEachWithPrefix(prefix []byte, consumer func([]byte, []byte)) error {
-	return this.forEach(append(this.prefix, prefix...), consumer)
+		// up on the first caller, check whether the version of the database is compatible
+		checkDatabaseVersion(dbDirClear)
+	})
+	return instance
 }
 
-func (this *prefixDb) ForEach(consumer func([]byte, []byte)) error {
-	return this.forEach(this.prefix, consumer)
+func CleanupGoShimmerBadgerInstance() {
```

Comment:
This function is not used. Should it be deleted?

Reply:
Made a background worker which calls it every 5 minute. See https://github.com/dgraph-io/badger#garbage-collection

Reply:
Where is it called from exactly?

Reply:
Can we somehow replace the `Println` with `log.Info` at least when the logger is initialized?

Reply:
Yeah, that's how I changed it, the logger is passed into the function.

Reply:
ah sorry, I thought you didn't change that.
Perfect then